### PR TITLE
Use Standard Dependency Workflow

### LINKED ISSUE #254: Implement Dependency RFC 0004 for  curl
Implement [Dependency Management RFC Phase 1](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md) for `curl`. Check out the RFC for more details and background.

### 1. Determine dependency source strategy
<details>
<summary>Background</summary>

When possible, dependencies should be used directly from upstream, rather than undergoing any additional compilation or modifications performed by Paketo-maintained code. For each dependency, the corresponding buildpack maintainer group will decide if the dependency can be used directly from upstream, and must identify the location from which the dependency will be pulled from. Some of the Paketo Java buildpacks perform directory stripping during the buildpack build process itself. This could be a viable alternative to performing directory modifications during the dependency management process for maintainers to consider.
</details>

- [x] Determine whether the dependency can be used directly from its upstream, rather than undergoing additional compilation/modification. 
* If using directly from upstream, note where dependencies will come from
* If compiled or modified, note the decision for this
- [x] Document decision in a language-family level RFC. This decision can be combined in the RFC with the decision made for other buildpacks in the language family.

### 2. Version retrieval and metadata generation code
Please refer to the [retrieval RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#1-version-retrieval-make-retrieve) on version retrieval for details around the API and background.

- [ ] Create retrieval code for the dependency
* Will live in the buildpack under: `dependency/retrieval/`
* Per [the RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#transfercombine-version-retrieval-and-metadata-generation), this will be a combination of (1) discovering new dependency versions based on the `buildpack.toml`, and (2) generating metadata for each new version.
* **Note** : Per [the RFC caveat](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#caveat-compiled-dependencies) - if the dependency is to be compiled, the `SHA256` and `URI` field from the metadata should be omitted in this step.

### 3. Compilation code (if needed)
If the dependency will be compiled/modified, then refer to the [compilation RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#2-compilation) for API details and background. 

- [ ] Create compilation action
* Will live in the buildpack under: `dependency/actions/compile/`

### 4. Dependency testing (optional)
It's up to maintainer discretion if the dependency will be tested. It's recommended to test the dependency if it's been compiled. Check out the [testing RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#3-test-dependency-make-test) for details.

- [ ] Add tests for dependency
* Will live in the buildpack under: `dependency/test.`

### 5. Makefile Setup
When using the generalized workflows for dependency management down the line, version retrieval and dependency testing will be executed via a Makefile in order to provide the workflow a standardized way to run the code, regardless of what language it was written in. Check out the [RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#overarching-plan) for what this should look like.

- [ ] Add Makefile
* Will live in the buildpack at: `dependency/Makefile`

### 6. Leveraging the new code
This issue serves to set up all the main logic for the dependency management. The work to actually leverage this code and migrate off of the dep-server will be completed in a separate issue once workflows and infrastructure is set up.

### PR DESCRIPTION
**Pull request by joshuatcasey**
## Summary
Use the standard dependency workflow. Resolves #254.

### CODE REVIEW

**Comment by joshuatcasey** [comment]
This PR should provide a generally equivalent result as `dep-server` as guided by https://github.com/paketo-buildpacks/rfcs/blob/main/text/ruby/0005-ruby-dependencies.md.

That being said, I'm really curious about our whole approach to this.

* Should we have a separate `curl` buildpack just for installing `curl`?
* Ruby is only supported on `base` and `full`. Full already has `curl` on the run image.
  * Should we check for the previous existence of `curl` on the path before installing `curl`?
  * Should we add `curl` to the base run image?
* How might we extend `curl` installation to the wildcard stack? The approach in this PR does not scale to arbitrary stacks. Should we move the compilation included in this PR into the buildpack?
* Remind me why `curl` is needed for `passenger`? It's not immediately obvious.

Note that the retrieval step is rather inefficient in that it downloads the source tarball three times:
* To verify the ASC signature
* To retrieve the checksum
* To search for licenses
I chose to optimize for simpler code rather than worrying about efficiency. Curl is only released a few times a year, so it's hardly a huge impact.

**Comment by joshuatcasey** [comment]
Note that this PR does include `jammy` compilation, but the buildpack will need to be marked with (A) jammy compatibility and (B) integration tested on `jammy`.

**Comment by joshuatcasey** [comment]
@sophiewigmore I'm tempted to recreate `curl@7.85.0` with the new dependency workflow so that we get immediate `jammy` support. 

We could throw away the recreated `bionic` dependency or swap it in for the current one.

What do you think?

**Comment by robdimsdale** [comment]
> * Should we have a separate `curl` buildpack just for installing `curl`?

I think that's overkill, as we haven't found any other buildpack that requires `curl` at runtime.

> * Ruby is only supported on `base` and `full`. Full already has `curl` on the run image.
>   
>   * Should we check for the previous existence of `curl` on the path before installing `curl`?

I'd rather not rely on `curl` being present in the stack, as in the general case it won't be present. The full stack is the exception - it provides everything for everyone for the purposes of getting started. It has too many packages to be a viable stack for production.

>   * Should we add `curl` to the base run image?

I'd prefer not to add `curl` to the base stack run image. It increases the surface area for security concerns significantly, and as mentioned above no other Paketo buildpack requires it at run-time.

> * How might we extend `curl` installation to the wildcard stack? The approach in this PR does not scale to arbitrary stacks. Should we move the compilation included in this PR into the buildpack?

Yes, in principle. If/when we add `*` stack support for the Ruby buildpack, we will have to add compilation logic for all the components (including MRI) to the relevant buildpacks. I don't think we have to solve that yet, though.

> * Remind me why `curl` is needed for `passenger`? It's not immediately obvious.

I don't remember and I haven't looked into it :)

> * Can we use a package manager to install curl? We could add a script that would execute on launch container startup, for example. `apt-get install curl -y` would have nice behavior in that it would do nothing if `curl` was already installed.

How would this work? The stacks do not have permission to execute arbitrary `apt` commands, by design, so we can't do this in the buildpack. Doing it in the stack is the same question as you ask above - should we add `curl` to the base stack run image?

**Comment by sophiewigmore** [comment]
I agree 100% with every response @robdimsdale  gave, however, I appreciate you asking these types of questions @joshuatcasey  to make sure we're doing the right thing.

Curl is needed to start passenger at launch- so it needs to be installed for base-stack purposes. We could probably get away with only installing curl if the stack doesn't include it, but it's probably easier to keep the logic as is

**Review by robdimsdale — APPROVED** [review]

**Comment by sophiewigmore** [comment]
> Note that the retrieval step is rather inefficient in that it downloads the source tarball three times
What's the main reason we couldn't reuse the same tarball?

**Comment by robdimsdale** [comment]
> We could probably get away with only installing curl if the stack doesn't include it, but it's probably easier to keep the logic as is

Yes, easier and also less logic to handle (less to test; less tech. debt to incur, etc)

File: dependency/Makefile
Comment (sophiewigmore) [inline dependency/Makefile:1]: I noticed there are no dependency tests, which I would hope for if we're compiling ourselves. Is there a reason you omitted them?

**Comment by sophiewigmore** [comment]
- This code works with the validator script (generates good metdata, compilation produces an artifact and checksum file)
- The compiled  bionic-target dependency works successfully in the buildpack
- I am wondering how you decided that the Jammy and Bionic variants need to be separately compiled. How do we know that the same dependency won't work for both?

**Comment by sophiewigmore** [comment]
Also if you add jammy support, then you need to add integration testing for jammy. It's not overly difficult to add, but I do not want to ship a dependency we haven't tested with the buildpacks.

**Comment by robdimsdale** [comment]
> I am wondering how you decided that the Jammy and Bionic variants need to be separately compiled. How do we know that the same dependency won't work for both?

I haven't explored this, but given we're compiling with the `--with-openssl` flag, I'd bet we're dynamically linking against `openssl`/`libssl` which we know is not ABI-compatible between bionic and jammy.

**Comment by sophiewigmore** [comment]
@robdimsdale ah I think you're right. I think the build I saw succeed must have passed due to some cached Docker volume magic.

**Comment by joshuatcasey** [comment]
I didn't test to see if a single dependency would work for both `bionic` and `jammy`, but I was aware that they have different versions of `libssl` so I didn't even try.

As for the dependency testing - I just forgot about that part. I see `dep-server` does have [some tests](https://github.com/paketo-buildpacks/dep-server/blob/main/actions/test-dependency/dependency-tests/curl/run) so I'll bring those over. I'll add that test now.

**Comment by joshuatcasey** [comment]
Created #291 to add `jammy` compatibility

**Review by robdimsdale — DISMISSED** [review]
This looks good to me.

I'm confused as to whether we still download the source tarball three times in the retrieval step - I couldn't see evidence of that but I don't see any changes either. I could be missing something though.

**Comment by joshuatcasey** [comment]
> I'm confused as to whether we still download the source tarball three times in the retrieval step - I couldn't see evidence of that but I don't see any changes either. I could be missing something though.

It still downloads three times. The source tarball is <5MB and it's only released a few times a year (maybe 5 or 6) so I didn't optimize.

**Comment by robdimsdale** [comment]
> It still downloads three times. The source tarball is <5MB and it's only released a few times a year (maybe 5 or 6) so I didn't optimize.

Yeah, I'm not concerned about optimizing either. I just couldn't see where we were doing three separate downloads. I see it now though:

1. In retrieval/main.go - `curlTarballPath, err := downloadToFile(sourceURL)`
2. In `compile/entrypoint` for Jammy: `curl "${upstream_tarball}" --output ...`
3. As above but for Bionic.

**Comment by robdimsdale** [comment]
@paketo-buildpacks/ruby-maintainers I'm happy with this but I'll leave it another couple of days for any of you all to take a look and share thoughts/feedback. If there's no feedback I'll merge it by the end of the week.

File: dependency/Makefile
Comment (ryanmoran) [inline dependency/Makefile:9]:
```suggestion
	go run ./retrieve \
	    --buildpack_toml_path=$(buildpackTomlPath) \
	    --output=$(output)
```

File: dependency/actions/compile/action.yml
Comment (ryanmoran) [inline dependency/actions/compile/action.yml:34]: Can we not just call the `docker` CLI directly since it exists on the runner?

**Comment by robdimsdale** [comment]
@sophiewigmore In addition to shipping jammy dependencies, we'd also have to declare the buildpack itself is jammy-compatible (via the top-level `[[stacks]]` key in buildpack.toml).

Phrased another way, until we add that in, the jammy dependencies are unused.

This gives us the option of adding the dependencies now and validating jammy support at a later date. I wouldn't want to wait too long, however, as adding the jammy dependencies will increase the size of the buildpack with no benefit until we add jammy support to the buildpack.

**Comment by sophiewigmore** [comment]
Thats reasonable! @robdimsdale

**Comment by sophiewigmore** [comment]
Will approve once Ryans comments are addressed

File: dependency/Makefile
Comment (joshuatcasey) [reply dependency/Makefile:9]: That's a good simplification. Will push commit after rebase.

**Comment by joshuatcasey** [comment]
@sophiewigmore @robdimsdale That's exactly the plan for #291 , to get that in place right after this PR is merged.

File: dependency/actions/compile/action.yml
Comment (joshuatcasey) [reply dependency/actions/compile/action.yml:34]: We sure can. Will push that commit.

**Pull request merged**
